### PR TITLE
Extract snapshotting logic to Snapshotter class

### PR DESCRIPTION
**Pull request by cbracken**
First step in eliminating code duplication between script snapshotting
(in FLX build) and AOT, assembly AOT snapshotting.

### CODE REVIEW

File: packages/flutter_tools/lib/src/base/build.dart [hunk 1]
```diff
@@ -6,8 +6,44 @@ import 'dart:async';
...
+
+class GenSnapshot {
+  const GenSnapshot();
+
+  Future<int> run({
+    @required TargetPlatform targetPlatform,
+    @required BuildMode buildMode,
+    @required String packagesPath,
+    @required String depfilePath,
+    @required Iterable<String> additionalArgs,
```
Comment (tvolkert) [inline packages/flutter_tools/lib/src/base/build.dart]: Or make it not required and default to `const <String>[]`?

File: packages/flutter_tools/lib/src/base/build.dart [hunk 2]
```diff
@@ -68,3 +104,107 @@ Future<Set<String>> readDepfile(String depfilePath) async {
...
+    _writeChecksum(outputSnapshotPath, depfilePath, mainPath, checksumsPath);
+    return 0;
+  }
+
+  Future<bool> _isBuildRequired(String outputSnapshotPath, String depfilePath, String mainPath, String checksumsPath) async {
+    final File checksumFile = fs.file(checksumsPath);
+    final File outputSnapshotFile = fs.file(outputSnapshotPath);
+    final File depfile = fs.file(depfilePath);
+    if (outputSnapshotFile.existsSync() && depfile.existsSync() && checksumFile.existsSync())
+    try {
```
Comment (tvolkert) [inline packages/flutter_tools/lib/src/base/build.dart]: Indentation.  Or maybe invert the `if` and return early.

**Review by tvolkert — APPROVED** [review]
LGTM

File: packages/flutter_tools/lib/src/base/build.dart (hunk 1 above)
Comment (cbracken) [inline packages/flutter_tools/lib/src/base/build.dart]: Good call. Done.

File: packages/flutter_tools/lib/src/base/build.dart (hunk 2 above)
Comment (cbracken) [inline packages/flutter_tools/lib/src/base/build.dart]: De Morgan's ftw. Done.

**Pull request merged**